File: src/main/scala/com/github/soniex2/notebetter/note/InstrumentBlock.java
package com.github.soniex2.notebetter.note;

import com.github.soniex2.notebetter.api.NoteBetterInstrument;
import com.google.common.base.Predicate;
import net.minecraft.block.state.IBlockState;

import java.util.ArrayList;
import java.util.List;

/**
 * @author soniex2
 */
public class InstrumentBlock {
    private List<Predicate<IBlockState>> predicates = new ArrayList<Predicate<IBlockState>>();
    private List<NoteBetterInstrument> instruments = new ArrayList<NoteBetterInstrument>();
    private NoteBetterInstrument fallback = null;

    public void addPredicate(Predicate<IBlockState> predicate, NoteBetterInstrument instrument) {
        predicates.add(predicate);
        instruments.add(instrument);
    }

    // badly named, but will do
    public void setDefaultInstrument(NoteBetterInstrument instrument) {
        this.fallback = instrument;
    }

    public NoteBetterInstrument get(IBlockState state) {
        for (int i = 0; i < predicates.size(); i++) {
            if (predicates.get(i).apply(state)) {
                return instruments.get(i);
            }
        }
        return fallback;
    }
}

File: src/main/scala/com/github/soniex2/notebetter/note/InstrumentBlock.scala
package com.github.soniex2.notebetter.note

import com.github.soniex2.notebetter.api.NoteBetterInstrument
import net.minecraft.block.state.IBlockState

import scala.collection.JavaConverters._
import scala.collection.immutable.{TreeMap, TreeSet}

/**
  * @author soniex2
  */
class InstrumentBlock(pred: List[(TreeMap[String, TreeSet[String]], NoteBetterInstrument)]) {

  def this() = this(List.empty)

  private var _predicates: List[(TreeMap[String, TreeSet[String]], NoteBetterInstrument)] = pred.sortBy(-_._1.size)
  //private var propTrie: PropTrie[Any with Comparable[Any], NoteBetterInstrument] = new PropTrie()

  def predicates = _predicates

  def addPredicate(predicate: TreeMap[String, TreeSet[String]], instrument: NoteBetterInstrument) {
    _predicates = ((predicate, instrument) :: _predicates).sortBy(-_._1.size)
  }

  def get(state: IBlockState): Option[NoteBetterInstrument] = {
    _predicates.find((v) => {
      state.getPropertyNames.asScala.flatMap((p) => v._1.get(p.getName).map((p, _))).forall((x) => x._2.contains(state.getValue(x._1).toString))
    }).map(_._2)
  }

  override def toString = {
    pred.toString()
  }
}